stellargraph/link_prediction.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from sklearn.pipeline import Pipeline
+from sklearn.linear_model import LogisticRegressionCV
+from sklearn.metrics import roc_auc_score
+from sklearn.preprocessing import StandardScaler
+import numpy as np
+import warnings
+
+
+_BINARY_OPERATORS = ["h", "avg", "l1", "l2"]
+
+
+def _operator_hadamard(u, v):
+    return u * v
+
+
+def _operator_avg(u, v):
+    return (u + v) / 2.0
+
+
+def _operator_l1(u, v):
+    return np.abs(u - v)
+
+
+def _operator_l2(u, v):
+    return (u - v) ** 2
+
+
+def _create_link_feature(pair, binary_operator):
+    u, v = pair
+    if binary_operator == "h":
+        return _operator_hadamard(u, v)
+    elif binary_operator == "avg":
+        return _operator_avg(u, v)
+    elif binary_operator == "l1":
+        return _operator_l1(u, v)
+    elif binary_operator == "l2":
+        return _operator_l2(u, v)
+    else:
+        raise ValueError(f"Unexpected binary operator: {binary_operator}")
+
+
+class LinkPredictionClassifier:
+    """
+    Link Prediction Classifier designed to be used with an existing node representation learner.
+
+    Args:
+        transform_node (callable): A function to transform node IDs into node features.
+        binary_operators (list of str, optional): Binary operators applied on node features to
+            produce the corresponding edge feature. Each binary operator will be used to train and
+            predict a separate classifier. By default, all binary operators are used.
+
+    """
+
+    def __init__(self, transform_node, binary_operators=None):
+        self.transform_node = transform_node
+        if binary_operators is not None:
+            self._binary_operators = binary_operators
+        else:
+            self._binary_operators = _BINARY_OPERATORS
+        self._classifiers = dict()
+
+    def _transform_node_pairs(self, link_examples, transform_node=None):
+        if transform_node is None:
+            transform_node = self.transform_node
+        return [
+            (transform_node(src), transform_node(dst)) for src, dst in link_examples
```

Comment:
Hmm, I guess then we have to enforce that the `transform_node` function being passed in by the user, supports the vectorised way of computing node features?

Maybe this makes more sense to do once we have node2vec implemented in our library, rather than using `gensim`'s word2vec, since the `transform_node` function will be coming from a trained word2vec model. I'd imagine currently this would just force the user to write a wrapper around the word2vec model to "appear" like it can do vectorised operations.

Reply:
Ah; I see. I guess it's fine as is, then (FWIW, vectorising an operation is fairly easy: `np.vectorize(some_func)`).